### ISSUE TITLE
I love your idea!

### ISSUE THREAD

**Issue by piranna** [streamich/jskernel#1]
Honestly, youar architecture seems to be almost perfect, congrats! :-D I decided to make NodeOS compatible with current apps, but if you could be able to make `jskernel` compatible with Node.js, I would be more than gladly to use it as the NodeOS kernel :-)

By the way, the architecture schemes are better done down-up, since they are stacks... ;-)

**Comment by streamich** [comment]
> By the way, the architecture schemes are better done down-up, since they are stacks... ;-)

That implies stacks grow upward ;-). But I agree they would look better if I drawn then _down-up_.

**Comment by piranna** [comment]
> That implies stacks grow upward ;-)

Yeah, the stacks always grow upward, specially if things are build _on top_ of other layers, and in last instance on top of the hardware ;-)

**Comment by ahmed-masud** [comment]
I concur! This is a fantastic idea, I'll be trying my best to contribute.

**Comment by 0bOne** [comment]
I am researching ideas for building a non-nix non-windows self-contained for-virtualization web/database server using the fewest languages possible. 

I gradually came to the conclusion that a good approach would be an exokernel based on something like Google's V8 engine. Node.js has too much functionality to match the 'absolute lightest' design constaints, of my project, but this project is very intriguing.   

I am interested in how far this project has come, whether it would be a suitable starting place to save me re-inventing the wheel.  

Do you think your project would be a good match? How far along is it?   Is it worth me contributing to this project or starting from scratch?

Thanks in advance
David

**Comment by streamich** [comment]
@dp68 I am continuing development in https://github.com/streamich/bamboo from time to time.

Things that are done:

  - I have runtime bindings for [Node, Duktape, and V8](https://github.com/streamich/bamboo/tree/master/packages) (see `bamboo-runtime-*`) those basically implement feautres of [`libsys`](https://github.com/streamich/libsys), that is all Bamboo needs to run. The most minimal version of Bamboo should need only `libsys.syscall()`.
  - [`libjs`](https://github.com/streamich/libjs) &mdash; is like `libc` in C, but in JavaScript.
  - I have [basic event loop with timers implemented](https://github.com/streamich/bamboo/blob/master/packages/bamboo-core/src/lib/eloop.ts) (which supports ref and un-ref queues and micro-tick queue), but it still needs self-pipe integration with thread pool.
  - I have [`asyscall` a lib that creates a thread pool](https://github.com/streamich/bamboo/tree/master/packages/asyscall) from JavaScript using [Assembler.js](https://github.com/streamich/ass-js), for asynchronous file system operations. It basically creates `process.asyscall()`.
  - I have [most of the file system module implemented](https://github.com/streamich/bamboo/blob/master/packages/bamboo-core/src/lib/fs.ts)
  - Other small things from node core are implemented.

Missing parts:

  - The biggest missing part is networking, I got basic prototypes working, but not much. Basically, `dns`, `udp`, `net`, `http` modules need to be created. When I will have time I will work on them.
  - Signals are not implemented but can be done relatively easily.
  - Many other things still need to be done...

**Comment by adminy** [comment]
Is there a super alpha? How would I be able to run it? Also what else is missing? Can I run this with qemu?

Thank you.
